src/vmm/src/vstate/vm.rs

```diff
@@ -181,16 +200,78 @@ impl Vm {
     pub fn register_memory_regions(
         &mut self,
         regions: Vec<GuestRegionMmap>,
+        userfault_bitmap_memfd: Option<&File>,
     ) -> Result<(), VmError> {
+        let addr = match userfault_bitmap_memfd {
+            Some(file) => {
+                // SAFETY: the arguments to mmap cannot cause any memory unsafety in the rust sense
+                let addr = unsafe {
+                    libc::mmap(
+                        std::ptr::null_mut(),
+                        usize::try_from(file.metadata().expect("Failed to get metadata").len())
+                            .expect("userfault bitmap file size is too large"),
+                        libc::PROT_WRITE,
+                        libc::MAP_SHARED,
+                        file.as_raw_fd(),
+                        0,
+                    )
+                };
+
+                if addr == libc::MAP_FAILED {
+                    panic!(
+                        "Failed to mmap userfault bitmap file: {}",
+                        std::io::Error::last_os_error()
+                    );
+                }
+
+                Some(addr as u64)
+            }
+            None => None,
+        };
```

Comment:
I'd prefer if this function took a pointer to the beginning of the userfault bitmap region, and then in the for loop below we do the pointer arithmetic to get the correct offset into it for the specific guest memory region.

But argh then this entire thing needs to be unsafe (well, `register_memory_region` already has to be given that it takes a u64 that we're happily interpreting as a pointer and passing to the kernel for doing reads and writes on). Maybe it can take a slice? We create the memfd, and before we pass it off to KVM for doing unholy things with it, having Rust slices to it should be fine.